rust/src/client.rs

```diff
@@ -0,0 +1,190 @@
+};
+use std::{sync::Arc, time::Duration};
+use thiserror::Error;
+use tokio::time;
+
+/// Client-side errors
+#[derive(Debug, Error)]
+pub enum ClientError {
+    /// Error starting the underlying [`Participant`]
+    #[error("failed to initialise participant: {0}")]
+    ParticipantInitErr(InitError),
+
+    /// Error from the underlying [`Participant`]
+    #[error("error arising from participant")]
+    ParticipantErr(PetError),
+
+    /// Error deserialising service data
+    #[error("failed to deserialise service data: {0}")]
+    DeserialiseErr(bincode::Error),
+
+    /// General client errors
+    #[error("unexpected client error")]
+    GeneralErr,
+}
+
+/// A client of the federated learning service
+///
+/// [`Client`] is responsible for communicating with the service, deserialising
+/// its messages and delegating their processing to the underlying
+/// [`Participant`].
+pub struct Client {
+    /// Handle to the federated learning [`Service`]
+    handle: Handle,
+
+    /// The underlying [`Participant`]
+    participant: Participant,
+
+    /// Interval to poll for service data
+    interval: time::Interval,
+
+    id: u32, // NOTE identifier for client for testing; may remove later
+}
+
+impl Client {
+    /// Create a new [`Client`]
+    ///
+    /// `period`: time period at which to poll for service data, in seconds.
+    /// Returns `Ok(client)` if [`Client`] `client` initialised successfully
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    pub fn new(period: u64) -> Result<Self, ClientError> {
+        let (handle, _events) = Handle::new();
+        let participant = Participant::new().map_err(ClientError::ParticipantInitErr)?;
+        Ok(Self {
+            handle,
+            participant,
+            interval: time::interval(Duration::from_secs(period)),
+            id: 0,
+        })
+    }
+
+    /// Create a new [`Client`] with ID (useful for testing)
+    ///
+    /// `period`: time period at which to poll for service data, in seconds.
+    /// `id`: an ID to assign to the [`Client`].
+    /// Returns `Ok(client)` if [`Client`] `client` initialised successfully
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    pub fn new_with_id(period: u64, handle: Handle, id: u32) -> Result<Self, ClientError> {
+        let participant = Participant::new().map_err(ClientError::ParticipantInitErr)?;
+        Ok(Self {
+            handle,
+            participant,
+            interval: time::interval(Duration::from_secs(period)),
+            id,
+        })
+    }
+
+    /// Start the [`Client`] loop
+    ///
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    /// NOTE in the future this may iterate only a fixed number of times, at the
+    /// end of which this returns `Ok(())`
+    pub async fn start(&mut self) -> Result<(), ClientError> {
+        loop {
+            // any error that bubbles up will finish off the client
+            self.during_round().await?;
+        }
```

Comment:
it still seems that the `update` and `sum2` messages will be sent multiple times per round even if everything goes properly on the coordinator side. basically a participant will await the `during_round()` which implies awaiting the `summer()` resp `updater()`. those two methods will most likely be completed before the coordinator declares the end of a round, hence the next loop iteration will poll the same round parameters and send the same messages over and over again.

a way to deal with this repetition would be check if the round parameters have changed (specifically the coordinator pk) and only call `summer()` resp `updater()` if this is the case. we can probably delay the implementation of such a check to another merge request (i already have a placeholder for such a pattern in the api mr) and still run the testing session with this version of the client.

Reply:
good point. in an older version, the idea was that each client would be in in sync from the start to the end of the round, but then this was changed to "return early" (the thinking was that it would make testing a bit easier). i agree that tracking the coordinator public key should resolve this.